File: webrtc_pixel_stream/ios/Classes/WebrtcPixelStreamPlugin.m
#import "WebrtcPixelStreamPlugin.h"
#import "FlutterRTCStreamingSink.h"

// Import the FlutterWebRTCPlugin header from the flutter_webrtc pod.
// This gives us access to +sharedSingleton and -trackForId:peerConnectionId:.
#import <flutter_webrtc/FlutterWebRTCPlugin.h>

// How many times to retry trackForId: before giving up (100ms apart)
static const int kMaxTrackLookupRetries = 20;

@implementation WebrtcPixelStreamPlugin {
  NSObject<FlutterBinaryMessenger> *_messenger;
  NSMutableDictionary<NSString *, FlutterRTCStreamingSink *> *_sinks;
  NSMutableDictionary<NSString *, RTCVideoTrack *> *_tracks;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FlutterMethodChannel *channel =
      [FlutterMethodChannel methodChannelWithName:@"webrtc_pixel_stream/control"
                                  binaryMessenger:[registrar messenger]];
  WebrtcPixelStreamPlugin *instance = [[WebrtcPixelStreamPlugin alloc] init];
  instance->_messenger = [registrar messenger];
  instance->_sinks = [NSMutableDictionary new];
  instance->_tracks = [NSMutableDictionary new];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
  if ([@"createPixelStream" isEqualToString:call.method]) {
    [self handleCreatePixelStream:call.arguments result:result];
  } else if ([@"disposePixelStream" isEqualToString:call.method]) {
    [self handleDisposePixelStream:call.arguments result:result];
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)handleCreatePixelStream:(NSDictionary *)args result:(FlutterResult)result {
  NSString *trackId = args[@"trackId"];
  id rawPcId = args[@"peerConnectionId"];

  if (!trackId || [trackId isKindOfClass:[NSNull class]] || trackId.length == 0) {
    result([FlutterError errorWithCode:@"INVALID_ARGS"
                               message:@"trackId is required"
                               details:nil]);
    return;
  }

  // Idempotency check
  if (_sinks[trackId]) {
    NSLog(@"[WebrtcPixelStream] Stream already exists for track: %@", trackId);
    NSString *channelName = [NSString stringWithFormat:@"webrtc_pixel_stream/frames/%@", trackId];
    result(@{@"channelName" : channelName});
    return;
  }

  // Sanitize peerConnectionId — treat NSNull as nil
  NSString *peerConnectionId = nil;
  if ([rawPcId isKindOfClass:[NSString class]] && [(NSString *)rawPcId length] > 0) {
    peerConnectionId = (NSString *)rawPcId;
  }

  FlutterWebRTCPlugin *webrtcPlugin = [FlutterWebRTCPlugin sharedSingleton];
  if (!webrtcPlugin) {
    result([FlutterError errorWithCode:@"NOT_READY"
                               message:@"FlutterWebRTCPlugin singleton is nil."
                               details:nil]);
    return;
  }

  NSLog(@"[WebrtcPixelStream] Creating sink for track: %@", trackId);

  // Phase 2: Extract Dart-allocated FFI buffer addresses
  NSNumber *addressA   = args[@"memoryAddressA"];
  NSNumber *addressB   = args[@"memoryAddressB"];
  NSNumber *sizeNumber = args[@"memorySize"];

  if (!addressA || !addressB || !sizeNumber ||
      [addressA isKindOfClass:[NSNull class]] ||
      [addressB isKindOfClass:[NSNull class]]) {
    result([FlutterError errorWithCode:@"INVALID_ARGS"
                               message:@"memoryAddressA, memoryAddressB and memorySize are required"
                               details:nil]);
    return;
  }

  uint8_t *sharedBufferA = (uint8_t *)[addressA unsignedLongLongValue];
  uint8_t *sharedBufferB = (uint8_t *)[addressB unsignedLongLongValue];
  size_t   bufferSize    = (size_t)[sizeNumber unsignedIntegerValue];

  NSLog(@"[WebrtcPixelStream] FFI buffers — A: %p  B: %p  size: %zu",
        sharedBufferA, sharedBufferB, bufferSize);

  FlutterRTCStreamingSink *sink =
      [[FlutterRTCStreamingSink alloc] initWithTrackId:trackId
                                             messenger:_messenger
                                         sharedBufferA:sharedBufferA
                                         sharedBufferB:sharedBufferB
                                            bufferSize:bufferSize];

  _sinks[trackId] = sink;

  // Return the channel name immediately so Dart can start listening.
  // The actual track attachment happens asynchronously with retries below.
  NSString *channelName = [NSString stringWithFormat:@"webrtc_pixel_stream/frames/%@", trackId];
  result(@{@"channelName" : channelName});

  // Delay the first lookup slightly — flutter_webrtc registers remote tracks
  // asynchronously and the track may not be visible to trackForId: yet.
  NSLog(@"[WebrtcPixelStream] Delaying renderer attachment for track: %@", trackId);
  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(150 * NSEC_PER_MSEC)),
                 dispatch_get_main_queue(), ^{
    [self attachRenderer:sink
                 trackId:trackId
      peerConnectionId:peerConnectionId
           webrtcPlugin:webrtcPlugin
            retryCount:0];
  });
}

/// Attempt to look up the video track and attach the renderer.
/// Retries up to kMaxTrackLookupRetries times (100ms apart) to handle
/// the race between our call and flutter_webrtc registering the remote track.
- (void)attachRenderer:(FlutterRTCStreamingSink *)sink
               trackId:(NSString *)trackId
    peerConnectionId:(NSString * _Nullable)peerConnectionId
         webrtcPlugin:(FlutterWebRTCPlugin *)webrtcPlugin
          retryCount:(int)retryCount {

  // If the sink was disposed while we were waiting, bail out.
  if (!self->_sinks[trackId]) {
    NSLog(@"[WebrtcPixelStream] Sink removed before attachment, aborting for track: %@", trackId);
    return;
  }

  RTCMediaStreamTrack *track = nil;

  // Guard the call inside @try/@catch. flutter_webrtc's trackForId:
  // can crash with -[NSNull stdString] if its internal peer-connection
  // dictionary contains NSNull placeholder entries (a known flutter_webrtc
  // bug when a connection is being negotiated).
  @try {
    track = [webrtcPlugin trackForId:trackId peerConnectionId:peerConnectionId];
  } @catch (NSException *exception) {
    NSLog(@"[WebrtcPixelStream] ⚠️ Exception in trackForId: (attempt %d): %@ – %@",
          retryCount + 1, exception.name, exception.reason);
    track = nil; // will retry below
  }

  if (track && [track isKindOfClass:[RTCVideoTrack class]]) {
    RTCVideoTrack *videoTrack = (RTCVideoTrack *)track;

    NSLog(@"[WebrtcPixelStream] Found video track: %@, readyState=%ld",
          trackId, (long)videoTrack.readyState);

    // CRITICAL: Only call addRenderer: when the track is live (readyState == 1).
    // Calling addRenderer: on a track with readyState=0 triggers abort() inside
    // WebRTC's C++ layer (via RTC_DCHECK) — this CANNOT be caught by @try/@catch
    // and will kill the process immediately.
    // RTCMediaStreamTrackStateLive = 1, RTCMediaStreamTrackStateEnded = 0
    if (videoTrack.readyState != RTCMediaStreamTrackStateLive) {
      NSLog(@"[WebrtcPixelStream] ⏳ Track not yet live (state=%ld), retrying (attempt %d/%d)...",
            (long)videoTrack.readyState, retryCount + 1, kMaxTrackLookupRetries);
      // Fall through to the retry block below
    } else {
      self->_tracks[trackId] = videoTrack;
      @try {
        [videoTrack addRenderer:sink];
        NSLog(@"[WebrtcPixelStream] ✅ Renderer attached to track: %@", trackId);
      } @catch (NSException *exception) {
        NSLog(@"[WebrtcPixelStream] ❌ Exception attaching renderer: %@ – %@",
              exception.name, exception.reason);
        [self->_tracks removeObjectForKey:trackId];
        [self->_sinks removeObjectForKey:trackId];
      }
      return;
    }
  }

  // Track not found yet — retry if we have attempts remaining
  if (retryCount < kMaxTrackLookupRetries) {
    NSLog(@"[WebrtcPixelStream] Track not found (attempt %d/%d), retrying in 100ms...",
          retryCount + 1, kMaxTrackLookupRetries);
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(100 * NSEC_PER_MSEC)),
                   dispatch_get_main_queue(), ^{
      [self attachRenderer:sink
                   trackId:trackId
        peerConnectionId:peerConnectionId
             webrtcPlugin:webrtcPlugin
              retryCount:retryCount + 1];
    });
  } else {
    NSLog(@"[WebrtcPixelStream] ❌ Gave up waiting for track: %@ after %d attempts",
          trackId, kMaxTrackLookupRetries);
    [self->_sinks removeObjectForKey:trackId];
  }
}

- (void)handleDisposePixelStream:(NSDictionary *)args result:(FlutterResult)result {
  id rawTrackId = args[@"trackId"];
  if (!rawTrackId || [rawTrackId isKindOfClass:[NSNull class]]) {
    result(nil);
    return;
  }
  NSString *trackId = (NSString *)rawTrackId;

  FlutterRTCStreamingSink *sink = _sinks[trackId];
  RTCVideoTrack *videoTrack = _tracks[trackId];

  if (sink && videoTrack) {
    @try {
      [videoTrack removeRenderer:sink];
    } @catch (NSException *e) {
      NSLog(@"[WebrtcPixelStream] Exception during removeRenderer: %@", e.reason);
    }
    [sink dispose];
  }

  [_sinks removeObjectForKey:trackId];
  [_tracks removeObjectForKey:trackId];
  result(nil);
}

@end

File: webrtc_pixel_stream/android/src/main/kotlin/com/camertronix/webrtc_pixel_stream/WebrtcPixelStreamPlugin.kt
package com.camertronix.webrtc_pixel_stream

import android.os.Handler
import android.os.Looper
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.webrtc.VideoTrack

/**
 * WebrtcPixelStreamPlugin
 * 
 * A lightweight companion plugin that streams raw BGRA pixel data from a
 * WebRTC video track via EventChannel.
 * 
 * This plugin contains zero WebRTC code of its own. It depends on
 * flutter_webrtc and accesses its StateProvider to attach streaming
 * renderers to any video track.
 */
class WebrtcPixelStreamPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var messenger: BinaryMessenger
    
    // We cache the methodCallHandler instance to bypass the StateProvider limitation
    private var cachedHandler: Any? = null
    
    private val sinks = mutableMapOf<String, FlutterRTCStreamingSink>()
    private val videoTracks = mutableMapOf<String, VideoTrack>()
    private val sinkToPeerConnectionId = mutableMapOf<String, String?>()
    
    companion object {
        private const val TAG = "WebrtcPixelStream"
        private const val MAX_RETRIES = 20
        private const val RETRY_DELAY_MS = 100L
        private const val INITIAL_DELAY_MS = 150L
    }
    
    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        Log.d(TAG, "onAttachedToEngine called")
        
        // 1. We must retrieve the FlutterWebRTCPlugin instance.
        // FlutterPlugin.FlutterPluginBinding does not expose `flutterEngine` publicly.
        // We reflect its `flutterEngine` private field to get the plugin registry.
        try {
            var engine: Any? = null
            for (field in binding.javaClass.declaredFields) {
                if (field.type.name == "io.flutter.embedding.engine.FlutterEngine") {
                    field.isAccessible = true
                    engine = field.get(binding)
                    break
                }
            }

            if (engine != null) {
                // engine is io.flutter.embedding.engine.FlutterEngine
                // Call engine.getPlugins().get(FlutterWebRTCPlugin.class)
                val getPluginsMethod = engine.javaClass.getMethod("getPlugins")
                val pluginRegistry = getPluginsMethod.invoke(engine)

                val getMethod = pluginRegistry.javaClass.getMethod("get", Class::class.java)
                val webrtcPluginClass = Class.forName("com.cloudwebrtc.webrtc.FlutterWebRTCPlugin")
                val webrtcPlugin = getMethod.invoke(pluginRegistry, webrtcPluginClass)

                if (webrtcPlugin != null) {
                    val handlerField = webrtcPluginClass.getDeclaredField("methodCallHandler")
                    handlerField.isAccessible = true
                    // Cache the actual handler (MethodCallHandlerImpl) instead of StateProvider
                    cachedHandler = handlerField.get(webrtcPlugin)
                    Log.d(TAG, "✅ MethodCallHandlerImpl accessed successfully via reflection")
                } else {
                    Log.w(TAG, "⚠️ FlutterWebRTCPlugin not found in engine plugin registry")
                }
            } else {
                Log.w(TAG, "⚠️ Could not find FlutterEngine in FlutterPluginBinding")
            }
        } catch (e: Exception) {
            Log.w(TAG, "⚠️ Reflection failed: Could not access WebRTC methodCallHandler: ${e.message}")
        }
        
        channel = MethodChannel(binding.binaryMessenger, "webrtc_pixel_stream/control")
        channel.setMethodCallHandler(this)
        messenger = binding.binaryMessenger
        
        Log.d(TAG, "Plugin initialized")
    }
    
    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        Log.d(TAG, "onDetachedFromEngine called")
        
        channel.setMethodCallHandler(null)
        
        // Clean up all sinks
        sinks.values.forEach { it.dispose() }
        sinks.clear()
        videoTracks.clear()
        sinkToPeerConnectionId.clear()
        
        Log.d(TAG, "Plugin detached")
    }
    
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "createPixelStream" -> handleCreatePixelStream(call.arguments as? Map<*, *>, result)
            "disposePixelStream" -> handleDisposePixelStream(call.arguments as? Map<*, *>, result)
            else -> result.notImplemented()
        }
    }
    
    private fun handleCreatePixelStream(args: Map<*, *>?, result: MethodChannel.Result) {
        if (args == null) {
            result.error("INVALID_ARGS", "Arguments required", null)
            return
        }
        
        val trackId = args["trackId"] as? String
        if (trackId == null || trackId.isEmpty()) {
            result.error("INVALID_ARGS", "trackId required", null)
            return
        }
        
        val sinkId = (args["sinkId"] as? String)?.takeIf { it.isNotEmpty() } ?: trackId
        
        val peerConnectionId = args["peerConnectionId"] as? String
        
        Log.d(TAG, "createPixelStream called - trackId: $trackId, sinkId: $sinkId, pcoId: $peerConnectionId")
        
        // Idempotency check
        if (sinks.containsKey(sinkId)) {
            Log.d(TAG, "Stream already exists for sink: $sinkId")
            result.success(mapOf("channelName" to "webrtc_pixel_stream/frames/$sinkId"))
            return
        }
        
        // Check if handler is available
        if (cachedHandler == null) {
            Log.e(TAG, "WebRTC handler not available - cannot access video tracks")
            result.error("NOT_READY", "WebRTC handler not ready. Ensure flutter_webrtc is initialized.", null)
            return
        }
        
        // Extract FFI buffer addresses (Phase B)
        val addressA = (args["memoryAddressA"] as? Number)?.toLong()
        val addressB = (args["memoryAddressB"] as? Number)?.toLong()
        val size = (args["memorySize"] as? Number)?.toLong()
        
        // Create sink with FFI support if addresses provided
        val sink = if (addressA != null && addressB != null && size != null) {
            Log.d(TAG, "Creating FFI sink - addressA: $addressA, addressB: $addressB, size: $size")
            FlutterRTCStreamingSink(sinkId, messenger, addressA, addressB, size)
        } else {
            Log.d(TAG, "Creating legacy sink (no FFI addresses)")
            FlutterRTCStreamingSink(sinkId, messenger)
        }
        
        sinks[sinkId] = sink
        // Ensure peerConnectionId is stored for tracking
        sinkToPeerConnectionId[sinkId] = peerConnectionId
        
        // Return channel name immediately so Dart can start listening
        result.success(mapOf("channelName" to "webrtc_pixel_stream/frames/$sinkId"))
        
        // Attach renderer asynchronously with retries (mirrors iOS behavior)
        Handler(Looper.getMainLooper()).postDelayed({
            attachRenderer(sink, trackId, peerConnectionId, sinkId, retryCount = 0)
        }, INITIAL_DELAY_MS)
    }
    
    private fun attachRenderer(
        sink: FlutterRTCStreamingSink,
        trackId: String,
        peerConnectionId: String?,
        sinkId: String,
        retryCount: Int
    ) {
        // Check if sink was disposed while waiting
        if (!sinks.containsKey(sinkId)) {
            Log.d(TAG, "Sink removed before attachment, aborting for sink: $sinkId")
            return
        }
        
        // Get track via reflection on MethodCallHandlerImpl.getTrackForId
        val track = getVideoTrack(trackId, peerConnectionId)
        
        if (track != null) {
            try {
                track.addSink(sink)
                videoTracks[sinkId] = track
                Log.d(TAG, "✅ Renderer attached to track: $trackId (sink: $sinkId)")
                return
            } catch (e: Exception) {
                Log.e(TAG, "❌ Exception attaching renderer to track: $trackId", e)
                sinks.remove(sinkId)
                return
            }
        }
        
        // Track not found yet — retry if we have attempts remaining
        if (retryCount < MAX_RETRIES) {
            Log.d(TAG, "Track not found (attempt ${retryCount + 1}/$MAX_RETRIES), retrying in ${RETRY_DELAY_MS}ms...")
            Handler(Looper.getMainLooper()).postDelayed({
                if (sinks.containsKey(sinkId)) {
                    attachRenderer(sink, trackId, peerConnectionId, sinkId, retryCount + 1)
                }
            }, RETRY_DELAY_MS)
        } else {
            Log.e(TAG, "❌ Gave up waiting for track: $trackId after $MAX_RETRIES attempts")
            sinks.remove(sinkId)
        }
    }
    
    // Private reflection helper to properly find remote and local tracks
    private fun getVideoTrack(trackId: String, peerConnectionId: String?): VideoTrack? {
        val handler = cachedHandler ?: return null
        return try {
            // Method signature: VideoTrack getTrackForId(String trackId, String peerConnectionId)
            val method = handler.javaClass.getDeclaredMethod(
                "getTrackForId", String::class.java, String::class.java
            )
            method.isAccessible = true
            method.invoke(handler, trackId, peerConnectionId) as? VideoTrack
        } catch (e: Exception) {
            Log.w(TAG, "getTrackForId reflection failed: ${e.message}")
            null
        }
    }
    
    private fun handleDisposePixelStream(args: Map<*, *>?, result: MethodChannel.Result) {
        val trackId = args?.get("trackId") as? String
        val sinkId = (args?.get("sinkId") as? String) ?: trackId
        
        Log.d(TAG, "disposePixelStream called - trackId: $trackId, sinkId: $sinkId")
        
        if (sinkId != null) {
            val sink = sinks.remove(sinkId)
            val track = videoTracks.remove(sinkId)
            
            if (sink != null && track != null) {
                try {
                    track.removeSink(sink)
                    Log.d(TAG, "✅ Renderer removed from track")
                } catch (e: Exception) {
                    Log.e(TAG, "Exception during removeSink: ${e.message}")
                }
                sink.dispose()
            }
        }
        
        result.success(null)
    }
}
